### PR TITLE
Create polycrystal diffusion material

### LINKED ISSUE #19643: Create a diffusion function material for polycrystals
## Reason

Polycrystal diffusivity is needed to activate various diffusion mechanisms within appropriate phases and interfaces in a Polycrystal structure. 

## Design

Create a diffusion function material that assigns different diffusivities to different phases/regions.  

## Impact

New capability. No impact on existing classes and tests.

### PR DESCRIPTION
**Pull request by SudiptaBiswas**
Closes #19643

### CODE REVIEW

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+{
+  Real SumEtaij = 0.0;
+  Real SumEtaj = 0.0;
+  for (unsigned int i = 0; i < _op_num; ++i)
+    for (unsigned int j = 0; j < _op_num; ++j)
+      if (j != i)
+      {
+        SumEtaij += 9.0 * (*_vals[i])[_qp] * (*_vals[i])[_qp] * (*_vals[j])[_qp] * (*_vals[j])[_qp];
+        SumEtaj += 2.0 * 9.0 * (*_vals[i])[_qp] * (*_vals[j])[_qp] * (*_vals[j])[_qp];
+      }
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:97]: Could be looped more efficiently without loss of clarity

```suggestion
  for (const auto i: make_range(_op_num))
    for (const auto j: make_range(i + 1, _op_num))
    {
      SumEtaij += 18.0 * (*_vals[i])[_qp] * (*_vals[i])[_qp] * (*_vals[j])[_qp] * (*_vals[j])[_qp];
      SumEtaj += 18.0 * ((*_vals[i])[_qp] * (*_vals[j])[_qp] * (*_vals[j])[_qp] + (*_vals[j])[_qp] * (*_vals[i])[_qp] * (*_vals[i])[_qp]);
    }
```

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  Real c = _c[_qp];
+  c = std::abs(c) > 1.0 ? 1.0 : (c < 0.0 ? std::abs(c) : c);
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:100]: why is this treating negative concentrations between -1 and 0 like that?

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  Real mc = 1.0 - c;
+
+  // Compute diffusion function
+  _diff[_qp] = _b_index * _diff_bulk * _hm[_qp] + _v_index * _diff_void * _hb[_qp] +
+               30.0 * _diff_surf * _s_index * c * c * mc * mc + _diff_gb * SumEtaij * _gb_index;
+
+  _dDdc[_qp] = _b_index * _diff_bulk * _dhmdc[_qp] + _v_index * _diff_void * _dhbdc[_qp] +
+               30.0 * _diff_surf * _s_index * (2.0 * c * mc * mc - 2.0 * c * c * mc);
+  for (MooseIndex(_op_num) op_index = 0; op_index < _op_num; ++op_index)
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:109]:
```suggestion
  for (const auto op_index: make_range(_op_num))
```

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    _hb(getMaterialProperty<Real>("void_switch")),
+    _hm(getMaterialProperty<Real>("solid_switch")),
+    _dhbdc(getMaterialPropertyDerivative<Real>("void_switch", _c_name)),
+    _dhmdc(getMaterialPropertyDerivative<Real>("solid_switch", _c_name)),
+    _diff_bulk(getParam<Real>("Dbulk")),
+    _diff_void(getParam<Real>("Dvoid")),
+    _diff_surf(getParam<Real>("Dsurf")),
+    _diff_gb(getParam<Real>("Dgb")),
+    _s_index(getParam<Real>("surfindex")),
+    _gb_index(getParam<Real>("gbindex")),
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:63]: I don't like the naming of these members. `_index` implies integer valued IMO, like `op_index` below. What those are, are weights. And they should be named accordingly.

File: modules/phase_field/include/materials/PolycrystalDiffusivity.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  const Real _gb_index;
+  const Real _b_index;
+  const Real _v_index;
+
+  const unsigned int _op_num;
+  std::vector<const VariableValue *> _vals;
+  std::vector<MaterialProperty<Real> *> _dDdv;
+
+  std::vector<const MaterialProperty<Real> *> _dhbdv;
+  std::vector<const MaterialProperty<Real> *> _dhmdv;
```
Comment (dschwen) [inline modules/phase_field/include/materials/PolycrystalDiffusivity.h:58]: Can you please add doxygen comments here?

File: modules/phase_field/include/materials/PolycrystalDiffusivity.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+/*                                                              */
+/*            (c) 2017 Battelle Energy Alliance, LLC            */
+/*                     ALL RIGHTS RESERVED                      */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*             Under Contract No. DE-AC07-05ID14517             */
+/*             With the U. S. Department of Energy              */
+/*                                                              */
+/*             See COPYRIGHT for full restrictions              */
+/****************************************************************/
```
Comment (dschwen) [inline modules/phase_field/include/materials/PolycrystalDiffusivity.h:13]: wrong header

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*             Under Contract No. DE-AC07-05ID14517             */
+/*             With the U. S. Department of Energy              */
+/*                                                              */
+/*             See COPYRIGHT for full restrictions              */
+/****************************************************************/
+
+#include "PolycrystalDiffusivity.h"
+#include "libmesh/quadrature.h"
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:16]: I don't think this header is required

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+/*                                                              */
+/*            (c) 2017 Battelle Energy Alliance, LLC            */
+/*                     ALL RIGHTS RESERVED                      */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*             Under Contract No. DE-AC07-05ID14517             */
+/*             With the U. S. Department of Energy              */
+/*                                                              */
+/*             See COPYRIGHT for full restrictions              */
+/****************************************************************/
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:13]: wrong header

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    _v_index(getParam<Real>("voidindex")),
+    _op_num(coupledComponents("v")),
+    _vals(_op_num),
+    _dDdv(_op_num),
+    _dhbdv(_op_num),
+    _dhmdv(_op_num)
+
+{
+  if (_op_num == 0)
+    mooseError("Model requires op_num > 0");
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:74]:
```suggestion
    paramError("op_num", "Model requires a non zero number of order parameters.");
```

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    for (unsigned int j = 0; j < _op_num; ++j)
+      if (j != i)
+      {
+        SumEtaij += 9.0 * (*_vals[i])[_qp] * (*_vals[i])[_qp] * (*_vals[j])[_qp] * (*_vals[j])[_qp];
+        SumEtaj += 2.0 * 9.0 * (*_vals[i])[_qp] * (*_vals[j])[_qp] * (*_vals[j])[_qp];
+      }
+
+  Real c = _c[_qp];
+  c = std::abs(c) > 1.0 ? 1.0 : (c < 0.0 ? std::abs(c) : c);
+  Real mc = 1.0 - c;
```
Comment (dschwen) [inline modules/phase_field/src/materials/PolycrystalDiffusivity.C:101]:
```suggestion
  const Real mc = 1.0 - c;
```

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C (hunk 2 above)
Comment (SudiptaBiswas) [reply modules/phase_field/src/materials/PolycrystalDiffusivity.C:100]: When `c` can vary from `-1` to `1` (when `c' is the order parameter for pore/bubble phase). This is done to treat both positive and negative values in the same manner.

File: modules/phase_field/include/materials/PolycrystalDiffusivity.h (hunk 5 above)
Comment (SudiptaBiswas) [reply modules/phase_field/include/materials/PolycrystalDiffusivity.h]: Added comments.

File: modules/phase_field/src/materials/PolycrystalDiffusivity.C (hunk 4 above)
Comment (SudiptaBiswas) [reply modules/phase_field/src/materials/PolycrystalDiffusivity.C]: Changed it.

**Review by dschwen — APPROVED** [review]

**Pull request merged**